### PR TITLE
Likelihood quadrature cleanup

### PR DESCRIPTION
**Pull request by st--**
**PR type:** cleanup

**Related issue(s)/PRs:** #1505, #1559, #1566

## Summary

**Proposed changes**

* Global constant for num_gauss_hermite_points default
* QuadratureLikelihood with dependency injection (in preparation for unifying with MonteCarloLikelihood)
* ScalarLikelihood as QuadratureLikelihood subclass (with _quadrature_dim/_quadrature_log_prob/_quadrature_reduction)
* HeteroskedasticTFPConditional: rename argument to `scale_transform` and remove undocumented scaling
* DeprecationWarning for deprecated code

## PR checklist

- [X] Code has type annotations
- [X] I ran the black formatter (`make format`)
- [x] I locally tested that the tests pass (`make check-all`)

### Release notes

**Fully backwards compatible:** yes / no

**If not, why is it worth breaking backwards compatibility:**

**Commit message (for release notes):**

* ...

### CODE REVIEW

File: gpflow/likelihoods/multilatent.py [hunk 1]
```diff
@@ -112,11 +110,9 @@ def __init__(
         def conditional_distribution(Fs) -> tfp.distributions.Distribution:
             tf.debugging.assert_equal(tf.shape(Fs)[-1], 2)
             loc = Fs[..., :1]
-            scale = transform(Fs[..., 1:] / 2)
+            scale = transform(Fs[..., 1:])
```
Comment (st--) [inline gpflow/likelihoods/multilatent.py:113]: @vdutor flagging this up as it's changing behaviour - I'm assuming you had the division by 2 so that the latent GP models the log-variance, but that's only true if the transform is actually Exp, so that seemed a bit weird.

File: gpflow/likelihoods/multilatent.py [hunk 2]
```diff
@@ -99,24 +97,23 @@ class HeteroskedasticTFPConditional(MultiLatentTFPConditional):
...
+        :param scale_transform: callable/bijector applied to the latent
+            function modelling the scale to ensure its positivity.
             Typically, `tf.exp` or `tf.softplus`, but can be any function f: R -> R^+.
         """
 
         def conditional_distribution(Fs) -> tfp.distributions.Distribution:
             tf.debugging.assert_equal(tf.shape(Fs)[-1], 2)
             loc = Fs[..., :1]
-            scale = transform(Fs[..., 1:] / 2)
+            scale = scale_transform(Fs[..., 1:])
```
Comment (st--) [inline gpflow/likelihoods/multilatent.py:114]: @vdutor flagging this up as it's mildly changing behaviour - I'm assuming you had the division by 2 so that the latent GP models the log-variance, but that's only true if the transform is actually Exp, so that seemed a bit weird for the general case here. (You can always include a scale in the transform itself!)

File: gpflow/likelihoods/multilatent.py [hunk 3]
```diff
@@ -99,24 +97,23 @@ class HeteroskedasticTFPConditional(MultiLatentTFPConditional):
     def __init__(
         self,
         distribution_class: Type[tfp.distributions.Distribution] = tfp.distributions.Normal,
-        transform: tfp.bijectors.Bijector = positive(base="exp"),
-        num_gauss_hermite_points: int = 21,
+        scale_transform: tfp.bijectors.Bijector = positive(base="exp"),
```
Comment (st--) [inline gpflow/likelihoods/multilatent.py:100]: Updated name to make clear that this should be a transform for the *scale*

File: gpflow/likelihoods/base.py [hunk 4]
```diff
@@ -366,15 +382,31 @@ class ScalarLikelihood(Likelihood):
...
         super().__init__(latent_dim=None, observation_dim=None, **kwargs)
-        self.num_gauss_hermite_points = 20
 
     @property
-    def num_gauss_hermite_points(self):
+    def num_gauss_hermite_points(self) -> int:
+        if not isinstance(self.quadrature, NDiagGHQuadrature):
+            raise TypeError(
+                "Can only query num_gauss_hermite_points if quadrature is a NDiagGHQuadrature instance"
+            )
```
Comment (gustavocmv) [inline gpflow/likelihoods/base.py:391]: maybe add some note like "quadrature-specific attributes can be accessed with likelihood.quadrature"?

File: gpflow/likelihoods/multilatent.py (hunk 2 above)
Comment (gustavocmv) [reply gpflow/likelihoods/multilatent.py:114]: tbh, I personally prefer to have the latent GP to log the log-scale, even though it is a little different from the HGP paper approach. Ofc, just a matter of taste.

Comment (vdutor) [reply gpflow/likelihoods/multilatent.py:114]: Happy for it to be the log-scale.

File: gpflow/likelihoods/base.py [hunk 5]
```diff
@@ -366,15 +382,36 @@ class ScalarLikelihood(Likelihood):
...
         return self.quadrature.n_gh
 
     @num_gauss_hermite_points.setter
-    def num_gauss_hermite_points(self, n_gh):
-        self.quadrature = NDiagGHQuadrature(1, n_gh)
+    def num_gauss_hermite_points(self, n_gh: int):
```
Comment (vdutor) [inline gpflow/likelihoods/base.py:400]: Can we not simply delete this instead of introducing deprecated code. I don't think many people changed the number of gh points on the fly any way.

File: gpflow/likelihoods/base.py [hunk 6]
```diff
@@ -289,62 +292,75 @@ def _variational_expectations(self, Fmu, Fvar, Y):
...
-                    "call a method which passes data to have it inferred."
-                )
-            with tf.init_scope():
-                self._quadrature = NDiagGHQuadrature(self.latent_dim, self.num_gauss_hermite_points)
-        return self._quadrature
+    def _quadrature_dim(self) -> int:
+        return self.latent_dim
 
-    def _predict_mean_and_var(self, Fmu, Fvar):
+    def _quadrature_log_prob(self, F, Y):
```
Comment (vdutor) [inline gpflow/likelihoods/base.py:313]: As `ScalarLikelihood` overwrites this method, does it make sense to have it private? As it seems to be important for inheriting classes, I think a docstring would be nice.

File: gpflow/likelihoods/base.py [hunk 7]
```diff
@@ -289,62 +292,75 @@ def _variational_expectations(self, Fmu, Fvar, Y):
...
+        return tf.expand_dims(self.log_prob(F, Y), axis=-1)
+
+    def _quadrature_reduction(self, v):
```
Comment (vdutor) [inline gpflow/likelihoods/base.py:316]: As `ScalarLikelihood` overwrites this method, does it make sense to have it private? As it seems to be important for inheriting classes, I think a docstring would be nice.

File: gpflow/likelihoods/multilatent.py [hunk 8]
```diff
@@ -99,24 +97,23 @@ class HeteroskedasticTFPConditional(MultiLatentTFPConditional):
...
+        **kwargs,
```
[the remark this answers is not in the document]
Comment (st--) [reply gpflow/likelihoods/multilatent.py:101]: I agree in principle though this is deprecating the `num_gauss_hermite_points` argument

File: tests/gpflow/likelihoods/test_heteroskedastic.py [hunk 9]
```diff
@@ -31,21 +31,18 @@ class Data:
...
-def likelihood() -> HeteroskedasticTFPConditional:
-    return HeteroskedasticTFPConditional(num_gauss_hermite_points=20)
-
-
-def test_analytic_mean_and_var(likelihood):
+def test_analytic_mean_and_var():
     """
     Test that quadrature computation used in HeteroskedasticTFPConditional
     of the predictive mean and variance is close to the analytical version, 
     which can be computed for the special case of N(y | mean=f1, variance=exp(f2)),
```
[the remark this answers is not in the document]
Comment (st--) [reply tests/gpflow/likelihoods/test_heteroskedastic.py:38]: thanks for the catch!

File: gpflow/likelihoods/base.py [hunk 10]
```diff
@@ -289,62 +292,75 @@ def _variational_expectations(self, Fmu, Fvar, Y):
...
+        return tf.squeeze(v, axis=-1)
```
[the remark this answers is not in the document]
Comment (st--) [reply gpflow/likelihoods/base.py:317]: Hm, i don't know. @gustavocmv ? I didn't want to mess about too much, just figure out how to get rid of the code duplication :thinking:

File: gpflow/likelihoods/base.py (hunk 5 above)
Comment (st--) [reply gpflow/likelihoods/base.py:400]: Then it'd be a breaking change though, and worst of all one users wouldn't even notice (they'd assign a different number, but it simply wouldn't have an effect anymore). If it's reasonably easy to add a backwards-compatible shim like in this case, that gives a deprecation warning, I think we should.

File: gpflow/likelihoods/base.py (hunk 7 above)
Comment (st--) [reply gpflow/likelihoods/base.py:316]: I don't think *other* classes would have to overwrite it, it's just because ScalarLikelihood behaves quite differently from the others ("broadcasting" over output dimensions and coupling latent_dim=observation_dim, without explicitly specifying either).

File: gpflow/likelihoods/base.py (hunk 6 above)
Comment (st--) [reply gpflow/likelihoods/base.py:313]: see below

File: gpflow/likelihoods/multilatent.py [hunk 11]
```diff
@@ -22,20 +22,18 @@
...
     r"""
     A Likelihood which assumes that a single dimensional observation is driven
     by multiple latent GPs.
+
+    Note that this implementation does not allow for taking into account
+    covariance between outputs.
     """
 
-    def __init__(self, latent_dim: int, *, num_gauss_hermite_points: int = 21):
+    def __init__(self, latent_dim: int, **kwargs):
```
[the remark this answers is not in the document]
Comment (st--) [reply gpflow/likelihoods/multilatent.py:34]: see above

File: gpflow/likelihoods/base.py (hunk 7 above)
Comment (vdutor) [reply gpflow/likelihoods/base.py:316]: Can you write this down in the code somewhere - maybe where you overwrite them.

File: gpflow/likelihoods/multilatent.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (vdutor) [reply gpflow/likelihoods/multilatent.py:101]: yes - that be nice

File: gpflow/likelihoods/base.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (vdutor) [reply gpflow/likelihoods/base.py:317]: If this is an easy change, I would be in favour of doing it. But it can be in a separate PR as well.

File: gpflow/likelihoods/multilatent.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (st--) [reply gpflow/likelihoods/multilatent.py:101]: On second thought, I'm not so sure about this. Apart from whether we want to do this for *all* likelihood classes (most just pass `**kwargs` up at the moment), it also means a _lot_ of changes if we decide to refactor it (though Pycharm might make that easier- do you know?)...

File: gpflow/likelihoods/base.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (gustavocmv) [reply gpflow/likelihoods/base.py:317]: I believe the main affected part is this one:

`tf.reduce_sum(fun(X, *args, **kwargs) * W, axis=0)`

Some functions used in quadrature e.g. `conditional_mean` will have output shape `[n_gh_total,] + batch_shape + [d,]`. The `W` variable is set to be broadcastable with it, so it has shape `[n_gh_total,] + [1,]*len(batch_shape) + [1,]`. I tried to have this (indirectly) covered in the docstrings as quoted below:
```
:param fun: Callable or Iterable of Callables that operates elementwise, with
            signature f(X, *args, **kwargs). Moreover, it must satisfy the shape-mapping:
                X shape: [N_quad_points, b1, b2, ..., bX, d],
                    usually [N_quad_points, N, d]
                f(X) shape: [N_quad_points, b1, b2, ...., bX, d'],
                    usually [N_quad_points, N, 1] or [N_quad_points, N, d]
            In most cases, f should only operate over the last dimension of X
```
If we want to support functions returning `[n_gh_total,] + batch_shape`, which is the case of `log_prob`, we'd need to check for it and squeeze `W`(or expand `fun(...)`) to match the shapes. I believe it was the behaviour of the old `ndiagquad` function, but I think it caused more confusion than helped (ofc my opinion, open to discuss this).

[the remark this answers is not in the document]
Comment (st--) [reply gpflow/likelihoods/base.py:317]: Thanks @gustavocmv ! I'm in favour of just leaving it as it is then; I've added these helpful explanations into the corresponding docstrings.

File: gpflow/likelihoods/base.py (hunk 7 above)
Comment (st--) [reply gpflow/likelihoods/base.py:330]: @vdutor @gustavocmv I've added docstrings, please check if a) they make sense and b) they satisfy your concerns!

File: gpflow/likelihoods/base.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (gustavocmv) [reply gpflow/likelihoods/base.py]: Glad to help @st-- 
@vdutor what do you think about it?

File: gpflow/likelihoods/base.py (hunk 7 above)
Comment (gustavocmv) [reply gpflow/likelihoods/base.py:335]: Checked them, lgtm.

**Review by vdutor — APPROVED** [review]
